File: app/src/main/java/com/apkupdater/adapter/UpdaterAdapter.java
package com.apkupdater.adapter;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.apkupdater.R;
import com.apkupdater.event.InstallAppEvent;
import com.apkupdater.event.SnackBarEvent;
import com.apkupdater.model.InstallStatus;
import com.apkupdater.model.LogMessage;
import com.apkupdater.model.Update;
import com.apkupdater.util.ColorUtil;
import com.apkupdater.util.DownloadUtil;
import com.apkupdater.util.GooglePlayUtil;
import com.apkupdater.util.LogUtil;
import com.apkupdater.util.MyBus;
import com.apkupdater.util.PixelConversion;
import com.apkupdater.util.SnackBarUtil;
import com.github.yeriomin.playstoreapi.AndroidAppDeliveryData;
import com.github.yeriomin.playstoreapi.GooglePlayException;
import com.squareup.otto.Subscribe;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.UiThread;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

@EBean
public class UpdaterAdapter
	extends RecyclerView.Adapter<UpdaterAdapter.UpdateViewHolder>
{
	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	private List<Update> mUpdates;
	private Context mContext;
	private RecyclerView mView;
    private Activity mActivity;

    @Bean
    MyBus mBus;

    @Bean
    LogUtil mLog;

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	public class UpdateViewHolder
		extends RecyclerView.ViewHolder
	{
		////////////////////////////////////////////////////////////////////////////////////////////////////////////////

		private View mView;
		private TextView mName;
		private TextView mPname;
		private TextView mVersion;
		private TextView mUrl;
		private ImageView mIcon;
		private Button mActionOneButton;
		private Button mActionTwoButton;
		private ImageView mIsBetaIcon;
        private ProgressBar mActionOneProgressBar;

		////////////////////////////////////////////////////////////////////////////////////////////////////////////////

		UpdateViewHolder(
			View view
		) {
			super(view);
			mView = view;
		}

		////////////////////////////////////////////////////////////////////////////////////////////////////////////////

		public void bind(
			final Update update
		) {
			// Get views
			mName = mView.findViewById(R.id.installed_app_name);
			mPname = mView.findViewById(R.id.installed_app_pname);
			mVersion = mView.findViewById(R.id.installed_app_version);
			mIcon = mView.findViewById(R.id.installed_app_icon);
			mIsBetaIcon = mView.findViewById(R.id.isbeta_icon);
			mUrl = mView.findViewById(R.id.update_url);
			mActionOneButton = mView.findViewById(R.id.action_one_button);
			mActionTwoButton = mView.findViewById(R.id.action_two_button);
            mActionOneProgressBar = mView.findViewById(R.id.action_one_progressbar);

			// Set values
			mName.setText(update.getName());
			mPname.setText(update.getPname());

			// Build version string with both old and new version
			String version = update.getVersion();
			if (update.getNewVersion() != null && !update.getNewVersion().isEmpty()) {
			    String newCode = update.getNewVersionCode() == 0 ? "?" : String.valueOf(update.getNewVersionCode());
				version += " (" + update.getVersionCode() + ") -> " + update.getNewVersion() + " (" + newCode  + ")";
			}

			mVersion.setText(version);

			// Build string for first action
			String action = "";
			String action2 = "";

			if (update.getUrl().contains("apkmirror.com")) {
				action = mContext.getString(R.string.action_apkmirror);
				//action2 = mContext.getString(R.string.action_evozi);
			} else if (update.getUrl().contains("uptodown.com")) {
				action = mContext.getString(R.string.action_uptodown);
                //action2 = mContext.getString(R.string.action_evozi);
			} else if (update.getUrl().contains("apkpure.com")) {
				action = mContext.getString(R.string.action_apkpure);
                //action2 = mContext.getString(R.string.action_evozi);
			} else if (update.getCookie() != null) {
			    action = mContext.getString(R.string.action_play);
            }
			mActionOneButton.setText(action);
            mActionTwoButton.setText(action2);

            if (action.equals(mContext.getString(R.string.action_play)) && update.getInstallStatus() != null) {
				mActionOneButton.setVisibility(View.VISIBLE);
				mActionOneProgressBar.setVisibility(View.INVISIBLE);
				mActionOneButton.setOnClickListener(null);
				if (update.getInstallStatus().getStatus() == InstallStatus.STATUS_INSTALL) {
					mActionOneButton.setText(R.string.action_play);
				} else if (update.getInstallStatus().getStatus() == InstallStatus.STATUS_INSTALLED) {
					mActionOneButton.setText(R.string.action_installed);
				} else if (update.getInstallStatus().getStatus() == InstallStatus.STATUS_INSTALLING) {
					mActionOneProgressBar.setVisibility(View.VISIBLE);
					mActionOneButton.setVisibility(View.INVISIBLE);
				}
			}

			// Action 1 listener
            if (!action.isEmpty()) {
                mActionOneButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        if (mActionOneButton.getText().equals(mContext.getString(R.string.action_play))) {

							// Check if we are already installing
							if (update.getInstallStatus() != null && update.getInstallStatus().getStatus() == InstallStatus.STATUS_INSTALLING) {
								return;
							} else {
								changeAppInstallStatusAndNotify(update, InstallStatus.STATUS_INSTALLING, 0, getAdapterPosition());
							}

							new Thread(new Runnable() {
                                @Override
                                public void run() {
                                    try {
                                        AndroidAppDeliveryData data = GooglePlayUtil.getAppDeliveryData(
                                        	GooglePlayUtil.getApi(mContext),
											update.getPname()
										);

                                        long id = DownloadUtil.downloadFile(
                                            mContext,
                                            data.getDownloadUrl(),
                                            data.getDownloadAuthCookie(0).getName() + "=" + data.getDownloadAuthCookie(0).getValue(),
                                            update.getPname() + " " + update.getNewVersionCode()
                                        );
										changeAppInstallStatusAndNotify(update, InstallStatus.STATUS_INSTALLING, id, getAdapterPosition());
                                    } catch (GooglePlayException gex) {
                                        SnackBarUtil.make(mActivity, String.valueOf(gex.getMessage()));
                                        mLog.log("UpdaterAdapter", String.valueOf(gex), LogMessage.SEVERITY_ERROR);
										changeAppInstallStatusAndNotify(update, InstallStatus.STATUS_INSTALL, 0, getAdapterPosition());
                                    } catch (Exception e) {
                                        SnackBarUtil.make(mActivity, "Error downloading.");
                                        mLog.log("UpdaterAdapter", String.valueOf(e), LogMessage.SEVERITY_ERROR);
										changeAppInstallStatusAndNotify(update, InstallStatus.STATUS_INSTALL, 0, getAdapterPosition());
                                    }
                                }
                            }).start();
                        } else {
                            DownloadUtil.LaunchBrowser(mContext, update.getUrl());
                        }
                    }
                });
            }

			// Action2 evozi listener
            if (!action2.isEmpty()) {
                mActionTwoButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(
                            "https://apps.evozi.com/apk-downloader/?id=" + update.getPname()
                        ));
                        mContext.startActivity(browserIntent);
                    }
                });
            }

			// Icon
			try {
				Drawable icon = mView.getContext().getPackageManager().getApplicationIcon(update.getPname());
				mIcon.setImageDrawable(icon);
			} catch (PackageManager.NameNotFoundException ignored) {}

			// Beta icon
			mIsBetaIcon.setVisibility(update.isBeta() ? View.VISIBLE : View.GONE);
			mIsBetaIcon.getBackground().setColorFilter(
				ColorUtil.getColorFromTheme(mContext.getTheme(), R.attr.colorAccent),
				android.graphics.PorterDuff.Mode.MULTIPLY
			);

			setTopMargin(0);
		}

		////////////////////////////////////////////////////////////////////////////////////////////////////////////////

        private void setTopMargin(
            int margin
        ) {
            ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) mView.getLayoutParams();
            params.topMargin = (int) PixelConversion.convertDpToPixel(margin, mContext);
        }

        ////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	public UpdaterAdapter(
		Context context
	) {
		mContext = context;
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public UpdateViewHolder onCreateViewHolder(
		ViewGroup parent,
		int viewType
	) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.updater_item, parent, false);
		return new UpdateViewHolder(v);
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public void onBindViewHolder(
		UpdateViewHolder holder,
		int position
	) {
		holder.bind(mUpdates.get(position));

        if (position == 0) {
		    holder.setTopMargin(8);
        }
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	@Override
	public int getItemCount(
	) {
		return mUpdates.size();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	public void addUpdate(
		Update update
	) {
		mUpdates.add(update);
		sort();

		int index = mUpdates.indexOf(update);

		notifyItemChanged(0);
		notifyItemInserted(index);
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	public void setUpdates(
		List<Update> updates
	) {
		mUpdates = updates;
		sort();
		notifyDataSetChanged();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	public int getCount(
	) {
		return mUpdates.size();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	private void sort(
	) {
		// Sorting non beta first, then alphabetically
		Collections.sort(mUpdates, new Comparator<Update>() {
			@Override
			public int compare(Update u1, Update u2) {
				return u1.isBeta() == u2.isBeta() ? u1.getName().compareToIgnoreCase(u2.getName()) : u1.isBeta() ? 1 : -1;
			}
		});
	}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    public void init(
        Activity activity,
        RecyclerView view,
        List<Update> updates
    ) {
        mActivity = activity;
        mView = view;
        setUpdates(updates);
        mBus.register(this);
    }

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	@UiThread
	protected void changeAppInstallStatusAndNotify(
		final Update app,
		int status,
		long id,
		final int pos
	) {
		if (app.getInstallStatus() != null) {
			app.getInstallStatus().setId(id);
			app.getInstallStatus().setStatus(status);
			notifyItemChanged(pos);
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	@Subscribe
	public void onInstallEvent(
		InstallAppEvent ev
	) {
		for (int i = 0; i < mUpdates.size(); i++) {
			Update app = mUpdates.get(i);
			if (app.getInstallStatus().getId() != 0 &&
				(app.getInstallStatus().getId() == ev.getId() || app.getPname().equals(ev.getPackageName())))
			{
				app.getInstallStatus().setId(0);
				if (ev.isSuccess()) {
					app.getInstallStatus().setStatus(InstallStatus.STATUS_INSTALLED);
					mBus.post(new SnackBarEvent(mContext.getString(R.string.install_success)));
				} else {
					// If the app is already set as installed, do nothing
					if (app.getInstallStatus().getStatus() == InstallStatus.STATUS_INSTALLING) {
						app.getInstallStatus().setStatus(InstallStatus.STATUS_INSTALL);
						mBus.post(new SnackBarEvent(mContext.getString(R.string.install_failure)));
					}
				}
				notifyItemChanged(i);

				// Delete file
				DownloadUtil.deleteDownloadedFile(mContext, app.getInstallStatus().getId());
			}
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

File: app/src/main/java/com/apkupdater/adapter/UpdaterAdapter.kt
package com.apkupdater.adapter

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import com.apkupdater.R
import com.apkupdater.event.InstallAppEvent
import com.apkupdater.event.RefreshUpdateTitle
import com.apkupdater.model.Update
import com.apkupdater.util.MyBus
import com.squareup.otto.Subscribe
import com.apkupdater.util.DownloadUtil
import com.apkupdater.event.SnackBarEvent
import com.apkupdater.model.IgnoreVersion
import com.apkupdater.model.InstallStatus
import com.apkupdater.model.MergedUpdate
import com.apkupdater.updater.UpdaterOptions
import com.apkupdater.util.InjektUtil
import uy.kohesive.injekt.api.get

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

class UpdaterAdapter
    : RecyclerView.Adapter<UpdaterViewHolder>
{
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    private var mUpdates : MutableList<Update>? = null
    private var mContext : Context? = null
    private var mView : RecyclerView? = null
	private val mBus : MyBus = InjektUtil.injekt?.get()!!
	private var mMergedUpdates : MutableList<MergedUpdate> = mutableListOf()

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    constructor(
        context : Context
    ) {
        mContext = context
        InjektUtil.addUpdaterAdapterSingleton(this)
        mBus.register(this)
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    fun init(
        view: RecyclerView,
        updates: MutableList<Update>
    ) {
        mView = view
        setUpdates(updates)
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    fun setUpdates(
        updates: MutableList<Update>
    ) {
        mUpdates = updates
        sort(true)
        notifyDataSetChanged()
    }

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	fun mergeUpdates(
		updates: MutableList<Update>
	) : MutableList<MergedUpdate> {
		val mergedUpdates : MutableList<MergedUpdate> = mutableListOf()
		val skip : MutableList<Pair<String, Int>> = mutableListOf()
		updates.forEach {
			if (!skip.contains(Pair(it.pname, it.newVersionCode))) {
				mergedUpdates.add(MergedUpdate(updates.filter {
					cit -> it.newVersionCode == cit.newVersionCode && it.pname == cit.pname
				}))
				skip.add(Pair(it.pname, it.newVersionCode))
			}
		}
		return mergedUpdates
	}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    fun sort(
	    b : Boolean
    ) {
	    // Filter and sort updates
	    mUpdates = sortUpdates(mContext as Context, mUpdates!!)
	    mMergedUpdates = mergeUpdates(mUpdates!!)

	    // Make sure we go to the start of the list
        if (b) {
	        mView?.layoutManager?.scrollToPosition(0)
        }
    }

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	companion object {
		fun sortUpdates(
			context: Context,
			updates: MutableList<Update>
		) : MutableList<Update> {
			val l = UpdaterOptions(context).ignoreVersionList
			return updates?.filter {
				val ignore = IgnoreVersion(it.pname, it.newVersion, it.newVersionCode)
				l.find {
					it.packageName == ignore.packageName
					&& it.versionName == ignore.versionName
					&& it.versionCode == ignore.versionCode
				} == null
			}?.sortedWith(compareBy(Update::isBeta).thenBy(Update::getName))?.toMutableList()
		}
	}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    fun addUpdate(
        update: Update
    ) {
        mUpdates?.add(update)
        sort(true)
        val index = mUpdates?.indexOf(update) as Int
	    notifyItemChanged(0)
        if (index >= 0) {
	        getIndexForUpdate(update, {index -> notifyItemInserted(index)})
        }
    }

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	fun removeUpdate(
		u : Update
	) {
		try {
			val i = mUpdates?.indexOf(u) as Int
			mUpdates?.removeAt(i)
			sort(false)
			getIndexForUpdate(u, {index -> notifyItemRemoved(index)})
			mBus.post(RefreshUpdateTitle())
		} catch (e : Exception) {}
	}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    override fun getItemCount(
    ): Int {
        return mMergedUpdates.size
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    override fun onBindViewHolder(
        holder: UpdaterViewHolder?,
        position: Int
    ) {
        holder?.bind(this, mMergedUpdates[position])

        if (position == 0) {
            holder?.setTopMargin(8)
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    override fun onCreateViewHolder(
        parent: ViewGroup?,
        viewType: Int
    ): UpdaterViewHolder {
        val v = LayoutInflater.from(parent?.context).inflate(R.layout.updater_item, parent, false)
        return UpdaterViewHolder(v)
    }

	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	fun getIndexForUpdate(
		u : Update,
		callback : (Int) -> Unit
	) {
		mMergedUpdates.forEachIndexed { i, it ->
			if (it.updateList.contains(u)) {
				callback(i)
				return
			}
		}
	}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @Subscribe
    fun onInstallAppEven(
        ev : InstallAppEvent
    ) {
	    mUpdates?.forEachIndexed { i, app ->
		    if (app.installStatus.id == ev.id) {
			    app.installStatus.id = 0
			    if (ev.isSuccess) {
				    app.installStatus.status = InstallStatus.STATUS_INSTALLED
				    mBus.post(SnackBarEvent(mContext?.getString(R.string.install_success)))
			    } else {
				    // If the app is already set as installed, do nothing
				    if (app.installStatus.status == InstallStatus.STATUS_INSTALLING) {
					    app.installStatus.status = InstallStatus.STATUS_INSTALL
					    mBus.post(SnackBarEvent(mContext?.getString(R.string.install_failure)))
				    }
			    }
			    getIndexForUpdate(app, {index -> notifyItemChanged(index)})
			    DownloadUtil.deleteDownloadedFile(mContext!!, app.installStatus.id)
		    }
	    }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////